### PR TITLE
election: make sure old `elec` will resign

### LINKED ISSUE #1212: evict leader can't take effect
## Bug Report

Please answer these questions before submitting your issue. Thanks!

1. What did you do? If possible, provide a recipe for reproducing the error.

in a test case [jenkins](https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/dm_ghpr_test/detail/dm_ghpr_test/7165/pipeline/), evicting leader failed.

2. What did you expect to see?

evict success

3. What did you see instead?

the DM-master which should be retired, gets the election key again so become leader.

[Untitled-2.log](https://github.com/pingcap/dm/files/5427934/Untitled-2.log)

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve? 
close #1212 

### What is changed and how it works?

if old election didn't resign, new election created with same session and key-prefix will always read that orphan "election key" by `Observe`, and can't clean it using `Resign`, unless TTL timeout (60s currently)

```
        elec := concurrency.NewElection(session, e.key)
	go func() {
		fmt.Println(elec.Campaign(ctx, e.key))
	}()
	ch := elec.Observe(ctx)
	info := <-ch
	fmt.Println("1", info)

	elec2 := concurrency.NewElection(session, e.key)
        // observe old election key
	ch2 := elec2.Observe(ctx)
	info3 := <-ch2
	fmt.Println("3", info3)
	fmt.Println(elec2.Resign(ctx))

	elec2 = concurrency.NewElection(session, e.key)
        // observe old election key
	ch2 = elec2.Observe(ctx)
	info3 = <-ch2
	fmt.Println("4", info3)
	fmt.Println(elec2.Resign(ctx))
```

### Check List 

Tests 

 - Unit test

Code changes

Side effects

Related changes

 - Need to cherry-pick to the release branch
 - Need to update release note
  fix a bug that sometimes can't evict leader of DM-master for a short time

### CODE REVIEW

**Comment by lance6716** [comment]
/run-all-tests

MySQL for unit test doesn't up

**Comment by lance6716** [comment]
```
[2020-10-25T11:38:24.874Z] ==================
[2020-10-25T11:38:24.874Z] WARNING: DATA RACE
[2020-10-25T11:38:24.874Z] Write at 0x00c0011bdfb8 by goroutine 252:
[2020-10-25T11:38:24.874Z]   go.etcd.io/etcd/clientv3/concurrency.(*Election).Resign()
[2020-10-25T11:38:24.874Z]       /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20191023171146-3cf2f69b5738/clientv3/concurrency/election.go:140 +0x5d8
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.(*Election).watchLeader.func1()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:365 +0x131
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.(*Election).watchLeader()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:404 +0x95a
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.(*Election).campaignLoop()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:325 +0x1a5e
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.NewElection.func1()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:143 +0x93
[2020-10-25T11:38:24.874Z] 
[2020-10-25T11:38:24.874Z] Previous write at 0x00c0011bdfb8 by goroutine 63:
[2020-10-25T11:38:24.874Z]   [failed to restore the stack]
[2020-10-25T11:38:24.874Z] 
[2020-10-25T11:38:24.874Z] Goroutine 252 (running) created at:
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.NewElection()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:141 +0x710
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/dm/master.(*Server).Start()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/master/server.go:214 +0xfbf
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/dm/master.(*testMaster).TestOperateSource()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/dm/master/server_test.go:1062 +0x8d1
[2020-10-25T11:38:24.874Z]   runtime.call32()
[2020-10-25T11:38:24.874Z]       /usr/local/go/src/runtime/asm_amd64.s:539 +0x3a
[2020-10-25T11:38:24.874Z]   reflect.Value.Call()
[2020-10-25T11:38:24.874Z]       /usr/local/go/src/reflect/value.go:321 +0xd3
[2020-10-25T11:38:24.874Z]   github.com/pingcap/check.(*suiteRunner).forkTest.func1()
[2020-10-25T11:38:24.874Z]       /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:850 +0x9aa
[2020-10-25T11:38:24.874Z]   github.com/pingcap/check.(*suiteRunner).forkCall.func1()
[2020-10-25T11:38:24.874Z]       /go/pkg/mod/github.com/pingcap/check@v0.0.0-20200212061837-5e12011dc712/check.go:739 +0x113
[2020-10-25T11:38:24.874Z] 
[2020-10-25T11:38:24.874Z] Goroutine 63 (finished) created at:
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.(*Election).campaignLoop()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:257 +0x1337
[2020-10-25T11:38:24.874Z]   github.com/pingcap/dm/pkg/election.NewElection.func1()
[2020-10-25T11:38:24.874Z]       /home/jenkins/agent/workspace/dm_ghpr_test/go/src/github.com/pingcap/dm/pkg/election/election.go:143 +0x93
[2020-10-25T11:38:24.874Z] ==================
```

didn't know why `pkg/election/election.go:143` and `pkg/election/election.go:141` caused a data race

File: pkg/election/election.go [hunk 1]
```diff
@@ -359,6 +361,14 @@ func (e *Election) watchLeader(ctx context.Context, session *concurrency.Session
 	e.campaignMu.Unlock()
 
 	defer func() {
+		e.l.Info("will try resign leader")
+		timeoutCtx, cancel := context.WithTimeout(context.Background(), time.Duration(e.sessionTTL)*time.Second)
+		if err := elec.Resign(timeoutCtx); err != nil {
+			e.l.Warn("fail to resign leader", zap.Stringer("current member", e.info), zap.Error(err))
+		} else {
+			e.l.Info("finish resign leader")
```
Comment (GMHDBJD) [inline pkg/election/election.go:369]: How about use debug, otherwise master will log it every TTL

File: pkg/election/election.go [hunk 2]
```diff
@@ -320,15 +320,16 @@ func (e *Election) campaignLoop(ctx context.Context, session *concurrency.Sessio
 			continue
 		}
 
+		// make sure `elec.Campaign` was exited (instead of rely on `<-eleObserveCh`), to avoid data race writing `elec`'s member
+		campaignWg.Wait()
+		cancel2()
```
Comment (GMHDBJD) [inline pkg/election/election.go:325]: Can we move it in front of L309 now?

**Comment by GMHDBJD** [comment]
So the issue is because ***ctx done***, then resign fail to remove the key?

**Comment by lance6716** [comment]
> So the issue is because _**ctx done**_, then resign fail to remove the key?

I'm not sure what happened in origin issue, only found there's an orphan election key and `Resign` didn't cover all branch. Maybe there're other cases cause this issue 😢

**Comment by lance6716** [comment]
Should we prepare to merge this PR before 2.0 GA, or wait better locating problem to prevent bring more bug?

File: pkg/election/election.go (hunk 2 above)
Comment (lance6716) [reply pkg/election/election.go:325]: I guess can't. For multi-leader election, non-leader's `elec.Campaign` will block thus can't `campaignWg.Done()`. If we move `campaignWg.Wait()` before above two if, behaviour will change to "failed campaign won't restart a new one". Not sure this is expected

**Comment by lance6716** [comment]
according to logs in https://internal.pingcap.net/idc-jenkins/blue/organizations/jenkins/dm_ghpr_test/detail/dm_ghpr_test/8507/pipeline/77/

I found there should be a txn deletes the old election KV

```
[2021-03-22T05:28:38.305Z] [2021/03/22 13:28:08.914 +08:00] [DEBUG] [interceptor.go:181] ["request stats"] [component="embed etcd"] ["start time"=2021/03/22 13:28:08.908 +08:00] ["time spent"=5.561244ms] [remote=127.0.0.1:58236] ["response type"=/etcdserverpb.KV/Txn] ["request count"=1] ["request size"=38] ["response count"=0] ["response size"=40] ["request content"="compare:<target:CREATE key:\"/dm-master/leader/12e07858679d7603\" create_revision:7 > success:<request_delete_range:<key:\"/dm-master/leader/12e07858679d7603\" > > failure:<>"]
```

before `fail to campaign`.

but for the buggy master that can't resign, there's no deleteOp

**Comment by lance6716** [comment]
can't reproduce 🤔

**Pull request closed without merging**